feature_extractor.py

```diff
@@ -0,0 +1,292 @@
+from torch.nn import functional as F
+from nle import nethack
+from torch import nn
+import torch
+
+
+def _step_to_range(delta, num_steps):
+    """Range of `num_steps` integers with distance `delta` centered around zero."""
+    return delta * torch.arange(-num_steps // 2, num_steps // 2)
+
+
+class Crop(nn.Module):
+    """Helper class for NetHackNet below."""
+
+    def __init__(self, height, width, height_target, width_target):
+        super(Crop, self).__init__()
+        self.width = width
+        self.height = height
+        self.width_target = width_target
+        self.height_target = height_target
+        width_grid = _step_to_range(2 / (self.width - 1), self.width_target)[
+            None, :
+        ].expand(self.height_target, -1)
+        height_grid = _step_to_range(2 / (self.height - 1), height_target)[
+            :, None
+        ].expand(-1, self.width_target)
+
+        # "clone" necessary, https://github.com/pytorch/pytorch/issues/34880
+        self.register_buffer("width_grid", width_grid.clone())
+        self.register_buffer("height_grid", height_grid.clone())
+
+    def forward(self, inputs, coordinates):
+        """Calculates centered crop around given x,y coordinates.
+        Args:
+           inputs [B x H x W]
+           coordinates [B x 2] x,y coordinates
+        Returns:
+           [B x H' x W'] inputs cropped and centered around x,y coordinates.
+        """
+        assert inputs.shape[1] == self.height
+        assert inputs.shape[2] == self.width
+
+        inputs = inputs[:, None, :, :].float()
+
+        x = coordinates[:, 0]
+        y = coordinates[:, 1]
+
+        x_shift = 2 / (self.width - 1) * (x.float() - self.width // 2)
+        y_shift = 2 / (self.height - 1) * (y.float() - self.height // 2)
+
+        grid = torch.stack(
+            [
+                self.width_grid[None, :, :] + x_shift[:, None, None],
+                self.height_grid[None, :, :] + y_shift[:, None, None],
+            ],
+            dim=3,
+        )
+
+        # TODO: only cast to int if original tensor was int
+        return (
+            torch.round(F.grid_sample(inputs, grid, align_corners=True))
```

Comment:
it would be nice to know what this is doing. Why do you need `grid_sample` as opposed to
`torch.narrow()` if all you're trying to express is `y = x[i0:i1, j0:j1]...`? What are you cropping and how?

Reply:
Below is the explanation I found in the "The NetHack Learning Environment" (NeurIPS 2020) paper. 
"We apply a ConvNet (red) to all visible glyph embeddings as well as another ConvNet (blue) to the 9×9 crop of glyphs around the agent to create a dedicated egocentric representation for improved generalization."
I spent a lot of time to understand what are they doing by grid_sample, but it does not make sense to me also... They are upsampling 9 by 9 grid to 21 by 79 grid, but why? And what's the meaning of interpolation regarding the fact that the numbers are glyphs, not RGB values (non-integer values will lose their meaning)?